### ISSUE TITLE
When using an `IModelFinalizingConvention` to add a shadow entity to the model, scripting migrations generates duplicate create table statements.

### ISSUE THREAD

**Issue by jeffreyabecker** [dotnet/efcore#31689]
When using a `ModelFinalizingConvention` to add a shadow entity to the model, scripting migrations generates duplicate create table statements.
[Full Repository Code:](https://github.com/jeffreyabecker/efcore-tabularhilo-wtf/tree/so-question/why-double-create)

I am trying to add a shadow entity to my model using a ModelFinalizing convention:
```
public class SequenceTableConvention : IModelFinalizingConvention
{
    public void ProcessModelFinalizing(IConventionModelBuilder modelBuilder, IConventionContext<IConventionModelBuilder> context)
    {
        var entityName = typeof(SequenceTable).FullName;
        var sequencesTableName = (string?)modelBuilder.Metadata.FindAnnotation(AnnotationNames.SequenceTableName)?.Value ?? SequenceTable.DefaultTableName;
        if (!modelBuilder.Metadata.GetEntityTypes().Any(e => e.Name == entityName))
        {
            var sequencesEntity = modelBuilder.Entity(entityName);

            sequencesEntity.Property(typeof(string), nameof(SequenceTable.SequenceName))
                .IsRequired(true)
                .HasMaxLength(255);
            sequencesEntity.Property(typeof(long), nameof(SequenceTable.CurrentValue))
                .IsRequired(true);
            sequencesEntity.Property(typeof(long), nameof(SequenceTable.IncrementBy))
                .IsRequired(true);
            sequencesEntity.PrimaryKey(new List<string> { nameof(SequenceTable.SequenceName) }, false);
            sequencesEntity.ToTable(sequencesTableName);

            var sequences = modelBuilder.Metadata.GetEntityTypes()
                .Where(HasHiLoSequence)
                .Select(GetSequenceInfo)
                .ToList();
            var entityType = (IMutableEntityType)sequencesEntity.Metadata;
            entityType.AddData(sequences);
        }
    }
...
}
```
 
when generating migrations with `dotnet ef migrations add InitialCreate` the Migration and ContextModelSnapshot code both contain statements declaring the table for the shadow entity a single time. However running `dotnet ef migrations script` produces the following output which includes duplicate create table statements for the shadow-entity's underlying table:
```
CREATE TABLE IF NOT EXISTS "__EFCoreHiLoSequences" (
    "SequenceName" TEXT NOT NULL CONSTRAINT "PK___EFCoreHiLoSequences" PRIMARY KEY,
    "CurrentValue" INTEGER NOT NULL,
    "IncrementBy" INTEGER NOT NULL
);
CREATE TABLE "__EFMigrationsHistory" (
    "MigrationId" TEXT NOT NULL CONSTRAINT "PK___EFMigrationsHistory" PRIMARY KEY,
    "ProductVersion" TEXT NOT NULL
);

BEGIN TRANSACTION;

CREATE TABLE "__EFCoreHiLoSequences" (
    "SequenceName" TEXT NOT NULL CONSTRAINT "PK___EFCoreHiLoSequences" PRIMARY KEY,
    "CurrentValue" INTEGER NOT NULL,
    "IncrementBy" INTEGER NOT NULL
);

CREATE TABLE "Blogs" (
    "Id" INTEGER NOT NULL CONSTRAINT "PK_Blogs" PRIMARY KEY AUTOINCREMENT,
    "Url" TEXT NOT NULL
);

CREATE TABLE "Posts" (
    "Id" INTEGER NOT NULL CONSTRAINT "PK_Posts" PRIMARY KEY AUTOINCREMENT,
    "Title" TEXT NOT NULL,
    "Content" TEXT NOT NULL,
    "BlogId" INTEGER NOT NULL,
    CONSTRAINT "FK_Posts_Blogs_BlogId" FOREIGN KEY ("BlogId") REFERENCES "Blogs" ("Id") ON DELETE CASCADE
);

INSERT INTO "__EFCoreHiLoSequences" ("SequenceName", "CurrentValue", "IncrementBy")
VALUES ('Blog', 1, 10);
INSERT INTO "__EFCoreHiLoSequences" ("SequenceName", "CurrentValue", "IncrementBy")
VALUES ('Post', 1, 10);

CREATE INDEX "IX_Posts_BlogId" ON "Posts" ("BlogId");

INSERT INTO "__EFMigrationsHistory" ("MigrationId", "ProductVersion")
VALUES ('20230911163442_InitialCreate', '7.0.10');

COMMIT;
```

EF Core version: 7.0.10
Database provider: Microsoft.EntityFrameworkCore.SqlServer, Micorsoft.EntityFrameworkCore.Sqlite
Target framework:  .NET 7.0
Operating system: Windows 11
IDE: Visual Studio 2022 17.6.3

**Comment by ajcvickers** [comment]
@jeffreyabecker 

> When using a ModelFinalizingConvention to add a shadow entity to the model,

I suspect that this will be very hard to make work because the code will need to make sure that the property is fully integrated with the model, which is normally done by reactive conventions (which can't run here, I don't think) and model finalizing conventions (which will likely already have run.) What is your goal for adding the shadow property and why are you doing this as a finalizing convention?

**Comment by jeffreyabecker** [comment]
Ultimately I've got a need to create a HiLo value generator which can be used across databases, including SQLite which we use in our CI/CD process. Other ORMs I've worked with manage this through a table. My hope with the convention was to get a create table statement + insert statements emitted during migration generation. As for why ModelFinalizingConvention -- it was the one that seemed to work / get my schema and records emitted into  the MyContextModelSnapshot.cs and MyMigration.Designer.cs files.

Currently I think that its something about the ModelFinalizingConvention running as part of the migration generation. If there's and easy way of just checking 'are we currently scripting out a migration' that would solve my problem.
